packages/ruleset/src/functions/circular-refs.js

```diff
@@ -0,0 +1,67 @@
+module.exports = function($ref, _opts, { path }) {
+  return checkForCircularRef($ref, path);
+};
+
+// This set is used to make sure that we warn about each distinct $ref value only once.
+const reportedRefValues = new Set();
+
+/**
+ * This function implements the "circular-refs" rule, which will check for
+ * $refs within the API definition that form cycles, either directly or indirectly.
+ * In this context, a circular ref (or cycle) would be a $ref to some sort of object (e.g. a schema)
+ * where traversing the referenced object's various sub-objects (e.g. schema properties,
+ * allOf/anyOf/oneOf lists, "additionalProperties", "items", etc.) leads us to a $ref that is a
+ * reference to the original referenced object.
+ * An example of a direct cycle would be a schema "Foo" that contains a property "foo" that is
+ * an instance of "Foo" itself".
+ * An example of an indirect cycle would be a "Foo" schema that contains property
+ * "bar" that is an instance of the "Bar" schema, and "Bar" contains a
+ * property "foo" that is an instance of the "Foo" schema.
+ * Any reference to either "Foo" or "Bar" will constitute a cyle.
+ *
+ * The way this rule is implemented is a bit odd so requires some explanation.
+ * The legacy "has_circular_refs" rule used both the resolved and unresolved API definition
+ * to perform its logic, where it trapsed through the entire API definition and identified
```

Comment:
This may be the first and last opportunity in my career to correct the spelling of the word "traipsed" in a code comment. 😂

```suggestion
 * to perform its logic, where it traipsed through the entire API definition and identified
```

Reply:
done